index_exprt now requires a type with subtype

kroening opened the pull request:
This commit adds a precondition to the constructor of `index_exprt` that
enforces that the type of the 'array' operand has a subtype.

A function is provided that offers the previous behavior to ease the
transition.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
index_exprt now requires a type with subtype

This commit adds a precondition to the constructor of index_exprt that
enforces that the type of the 'array' operand has a subtype.

tautschnig commented on src/analyses/variable-sensitivity/write_stack_entry.cpp:
```diff
@@ -71,7 +71,7 @@ exprt offset_entryt::get_access_expr() const
   // meaning that we don't know what the type is at this point, as the
   // array part will be filled in later.
-  return index_exprt(nil_exprt(), offset->to_constant());
+  return index_exprt::unchecked(nil_exprt(), offset->to_constant());
```
While intentional as per the comment, is there perhaps a way of rewriting this code?

kroening replied:
Sure there is; I tried and gave up.

tautschnig replied:
I have added a commit doing this rewrite.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -1416,12 +1416,19 @@ class index_exprt:public binary_exprt
   {
   }
 
+  // legacy constructor, which will be removed eventually
+  static index_exprt unchecked(exprt array, exprt index)
+  {
+    auto type = array.type().has_subtype()
+                  ? to_type_with_subtype(array.type()).subtype()
+                  : typet(ID_nil);
+    return index_exprt(std::move(array), std::move(index), std::move(type));
+  }
+
```
Should we mark it as deprecated right away? We need some way to track this.

kroening replied:
The build fails when you call a deprecated method.

tautschnig replied:
It is now no longer needed.

kroening replied:
ok, dropped

tautschnig commented on unit/solvers/smt2_incremental/object_tracking.cpp:
```diff
@@ -27,11 +27,6 @@ TEST_CASE("find_object_base_expression", "[core][smt2_incremental]")
     rowt{
       "Address of struct member",
       {member_exprt{object_base, "baz", unsignedbv_typet{8}}, pointer_type}},
-    rowt{
-      "Address of index of struct member",
-      {index_exprt{
-         member_exprt{object_base, "baz", unsignedbv_typet{8}}, index},
-       pointer_type}},
```
@NlightNFotis @thomasspriggs Could you please review this code (and the surrounding lines) to clarify what is meant to be going on here? Multiple type-inconsistent expressions appear to be formed here.

thomasspriggs replied:
`GENERATE_REF` is being used to generate a set of tests where the `find_object_base_expression` function is called with various expressions and is expected to return `object_base`. The `find_object_base_expression` function is analogous to part of the implementation of `pointer_logict::add_object`. The unit test currently works because the `find_object_base_expression` function does not currently depend on the `.type()` of the expressions or their sub-expressions and because the constructors of `index_exprt` and `member_exprt` do not check the type of `object_base`.

If you change the currently deleted lines of code to the following, then it should make the `index_exprt` sufficiently type-consistent to work with this PR -
```
    rowt{
      "Address of index of struct member",
      {index_exprt{
         member_exprt{object_base, "baz", base_type}, index},
       pointer_type}},
```

I acknowledge that this does not make the instances of `member_exprt` type-consistent. They could be made type-consistent as well but this would require expanding the test with differently typed base_object expressions for each generated test, using various `struct_typet`s where appropriate.

tautschnig committed:
Variable sensitivity: do not construct incomplete index_exprt

Instead of requiring an implicit sequence of method invocations make
explicit when the result is just an offset and when it is a full
expression.
